## B2_김대호_통계2일차(0511)
### 1. 정규분포의 확률 계산
#### (1) 1리터를 가지고 9km 이상 15km 이하를 달릴 확률
#### (2) 1리터를 가지고 15km 이상을 달릴 확률

In [3]:
import numpy as np
from scipy import stats
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

mu = 12
sigma = 3
x1 = 9
x2 = 15

prob1 = stats.norm.cdf(x1, mu, sigma)

print("{0}km 이상 {1}km 이하를 달릴 확률은 {2:.1f}에 해당함".format(x1, x2, prob2 - prob1))
print("{0}km 이상을 달릴 확률은 {1:.1f}에 해당함".format(x2, (1 - prob2)))

NameError: name 'prob2' is not defined

In [9]:
import numpy as np
from scipy import stats
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

ds_anova = pd.read_csv("./시험_예제/전표처리시간.csv", engine="python")

mu = 12
sigma = 3
x1 = 35

prob1 = stats.norm.cdf(x1, mu, sigma)
prob1

# print("{0}km 이상 {1}km 이하를 달릴 확률은 {2:.1f}에 해당함".format(x1, x2, prob2 - prob1))
# print("{0}km 이상을 달릴 확률은 {1:.1f}에 해당함".format(x2, (1 - prob2)))

0.9999999999999912

### 자동차 1리터당 주행거리 평균은 12km, 표준편차는 3km인 정규분포이다.
#### 정규분포의 그래프의 면적은 확률을 의미한다.
1. stats.norm.cdf 를 이용해서 자동차가 1리터로 15km 이하를 달릴 확률을 구한다. 약 0.16가 나온다
   stats.norm.cdf 를 이용해서 자동차가 1리터로 9km 이하를 달릴 확률을 구한다. 약 0.84가 나온다
   15km 이하를 달릴 확률에서 9km 이하를 달릴 확률을 빼면 9km 이상 15km 이하를 달릴 확률을 구할 수 있다.


2. stats.norm.cdf 를 이용해서 자동차가 1리터로 15km 이하를 달릴 확률을 구한다. 약 0.16가 나온다
   자동차가 달릴 전체 확률인 1에서 15km 이하로 달릴 확률을 빼면 15km 이상을 달릴 확률을 구할 수 있다.
   약 0.2가 나온다.
   
### 결론 : 주행거리가 길어질수록 자동차가 1리터로 갈 확률이 줄어든다.

## 2. 포아송 분포
### 일정한 시간 또는 공간에서 발생하는 성공횟수에 대한 이산확률분포
#### 한 상점 (일정한 공간) 에서 평균 1시간에 3명씩 포아송분포로 손님이 도착한다.

#### 10시에서 11시 사이에 (일정한 시간)에 다음이 발생할 확률을 구하시오

1) 2명의 손님이 도착

2) 1명 이하의 손님이 도착

3) 3명 이상의 손님이 도착

In [55]:
mu = 3
prob = stats.poisson.pmf(2, mu)
cdf_prob1 = stats.poisson.cdf(1, mu)
cdf_prob2 = stats.poisson.cdf(3, mu)

print("10시에서 11시 사이에 {0}명의 손님이 도착할 확률: {1:.1f}".format(2, prob))
print("10시에서 11시 사이에 {0}명 이하의 손님이 도착할 확률: {1:.1f}".format(1, cdf_prob1))
print("10시에서 11시 사이에 {0}명 이상의 손님이 도착할 확률: {1:.1f}".format(3, cdf_prob2))

10시에서 11시 사이에 2명의 손님이 도착할 확률: 0.2
10시에서 11시 사이에 1명 이하의 손님이 도착할 확률: 0.2
10시에서 11시 사이에 3명 이상의 손님이 도착할 확률: 0.6


## 1시간에 평균 3명

### 확률밀도함수 Pr(X = x) 이용

### 2명의 손님이 도착할 확률 = pmf(2, mu)
### 1명 이하의 손님이 도착할 확률 = cdf(1, mu)
### 3명 이상의 손님이 도착할 확률 = 1 - (3명 이하의 손님이 도착할 확률 cdf(3,mu))

## 3. 정규성 검정

### 정규성 검정결과 (95% 신뢰수준 사용시) p-value가 0.05보다 크면 정규, 0.05보다 작으면 비정규

####  Mat_A의 p-value > 0.05 이므로 정규
####  Mat_B의 p-value > 0.05 이므로 정규

In [54]:
from scipy.stats import shapiro
ds_mat = pd.read_csv("2sample_t_test.csv", engine="python")
mat_a = ds_mat.dropna()
mat_a = mat_a['Mat_A']

shapiro(mat_a)

ShapiroResult(statistic=0.9062636494636536, pvalue=0.2562994360923767)

In [44]:
from scipy.stats import shapiro
ds_mat = pd.read_csv("2sample_t_test.csv", engine="python")
mat_b = ds_mat.dropna()
mat_b = mat_b['Mat_B']

shapiro(mat_b)

ShapiroResult(statistic=0.9729753732681274, pvalue=0.9169777035713196)

## 4. 모평균 추정 (모표준편차 아는 경우)
### 모표준편차는 3.8이다.
### 정규분포 구간 추정 함수를 실행해 95% 신뢰도와 3.8의 표준편차를 갖는 모평균을 추정한다.

In [56]:
df = pd.DataFrame({"sample": [18, 18, 20, 21, 20, 23, 19, 18, 17, 21, 22, 20, 20, 21, 20, 19, 19, 18, 17, 19]})
df.head()

lower, upper = stats.norm.interval(0.95, loc = np.mean(df), scale=0.38/np.sqrt(40))
print("신뢰구간: ({0},{1})".format(lower.round(2), upper.round(2)))

신뢰구간: ([19.38],[19.62])


### 95% 신뢰 구간은 19.38 ~ 19.62 으로 추정한다

## 5. 모평균 추정(모표준편차 모르는 경우)
### 평균 인장강도 = 70 kg/mm^2 기대

#### 모평균값의 95% 신뢰구간을 추정

In [50]:
df = pd.DataFrame({'sample':[73, 71, 74, 69, 70, 73, 70, 68, 75, 72, 70, 72, 73, 70, 70, 72, 71, 70, 75, 72]})
    
lower, upper = stats.t.interval(0.95, len(df) - 1, loc = np.mean(df), scale = stats.sem(df))
print("신뢰구간: ({0},{1})".format(lower.round(2), upper.round(2)))

신뢰구간: ([70.6],[72.4])


### 95% 신뢰구간은 70.6 ~ 72.4 으로 추정한다.